### PR TITLE
internal/contour: Initial implementation of healthchecks / lb algo

### LINKED ISSUE #442: [IngressRoute] Healthcheck endpoints
Health checks are pec'd out as well in the IngressRoute design doc (https://github.com/heptio/contour/blob/master/design/ingressroute-design.md#healthcheck). I think the implementation of this is similar to #415 where HealthChecks are applied to clusters and we need to think about how to roll this into Contour's design.

### PR DESCRIPTION
**Pull request by stevesloka**
Fixes #442  & #415 by implementing (optional) health checks per service defined in an IngressRoute as well as lb algorithm

Signed-off-by: Steve Sloka <steves@heptio.com>

### CODE REVIEW

File: internal/contour/cluster.go [hunk 1]
```diff
@@ -25,9 +25,21 @@ import (
...
 	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	envoy_api_v2_core4 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	google_protobuf "github.com/gogo/protobuf/types"
+	ingressroutev1 "github.com/heptio/contour/apis/contour/v1beta1"
 	"github.com/heptio/contour/internal/dag"
 )
 
+const (
+	// Default healthcheck / lb algorithm values
+	hcTimeout            = int64(2)
```
Comment (davecheney) [inline internal/contour/cluster.go]: You can drop the cast here, leaving them as untyped numeric constants, the compiler will figure it out.

File: internal/contour/cluster.go [hunk 2]
```diff
@@ -107,6 +119,7 @@ func (v *clusterVisitor) Visit() map[string]*v2.Cluster {
 }
 
 func (v *clusterVisitor) visit(vertex dag.Vertex) {
+
```
Comment (davecheney) [inline internal/contour/cluster.go]: delete

File: internal/contour/cluster.go [hunk 3]
```diff
@@ -160,6 +178,70 @@ func (v *clusterVisitor) edscluster(svc *dag.Service) {
...
+		return v2.Cluster_ROUND_ROBIN
+	case "WeightedLeastRequest":
+		return v2.Cluster_LEAST_REQUEST
+	case "RingHash":
+		return v2.Cluster_RING_HASH
+	case "Maglev":
+		return v2.Cluster_MAGLEV
+	case "Random":
+		return v2.Cluster_RANDOM
+	}
```
Comment (davecheney) [inline internal/contour/cluster.go]: You can write this a bit more compactly like this
```
switch switch lbStrategy {
case "WeightedLeastRequest":
	return v2.Cluster_LEAST_REQUEST
case "RingHash":
	return v2.Cluster_RING_HASH
case "Maglev":
	return v2.Cluster_MAGLEV
case "Random":
	return v2.Cluster_RANDOM
default:
	return v2.Cluster_ROUND_ROBIN
}
```

File: internal/contour/cluster_test.go [hunk 4]
```diff
@@ -20,6 +20,8 @@ import (
 
 	"github.com/envoyproxy/go-control-plane/envoy/api/v2"
 	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	envoy_api_v2_core4 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
```
Comment (davecheney) [inline internal/contour/cluster_test.go]: why is this package imported twice?

File: internal/contour/cluster_test.go [hunk 5]
```diff
@@ -240,6 +242,383 @@ func TestClusterVisit(t *testing.T) {
...
+					Name: "default/backend/80",
+					Type: v2.Cluster_EDS,
+					EdsClusterConfig: &v2.Cluster_EdsClusterConfig{
+						EdsConfig:   apiconfigsource("contour"), // hard coded by initconfig
+						ServiceName: "default/backend/http",
+					},
+					ConnectTimeout: 250 * time.Millisecond,
+					LbPolicy:       v2.Cluster_ROUND_ROBIN,
+					HealthChecks: []*envoy_api_v2_core4.HealthCheck{{
+						Timeout: &google_protobuf.Duration{
```
Comment (davecheney) [inline internal/contour/cluster_test.go]: Hmm, something is weird here with the grpc types, look at ConnectTimeout: that takes a time.Duration, why do these fields take a protobuf.duration? is it because they are optional? (i've seen other stuff in envoy where a pointer to zero means infinity but a nil pointer means -- effectively -- zero.

File: internal/contour/cluster_test.go [hunk 6]
```diff
@@ -240,6 +242,383 @@ func TestClusterVisit(t *testing.T) {
...
```
Comment (davecheney) [inline internal/contour/cluster_test.go]: please add a little helper
```
func duration(d time.Duration) &google_protobuf.Duration {
   ...
}
```

File: internal/contour/cluster_test.go [hunk 7]
```diff
@@ -20,6 +20,8 @@ import (
 
 	"github.com/envoyproxy/go-control-plane/envoy/api/v2"
 	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	envoy_api_v2_core4 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	google_protobuf "github.com/gogo/protobuf/types"
```
Comment (davecheney) [inline internal/contour/cluster_test.go]: why is this import renamed, `types` shouldn't conflict with anything in the file.

File: internal/dag/dag.go [hunk 8]
```diff
@@ -410,7 +410,7 @@ func (d *DAG) processIngressRoute(ir *ingressroutev1.IngressRoute, prefixMatch s
 			for _, s := range route.Services {
 				m := meta{name: s.Name, namespace: ir.Namespace}
 				if svc := service(m, intstr.FromInt(s.Port)); svc != nil {
-					r.addService(svc)
+					r.addService(svc, s.HealthCheck, s.Strategy)
```
Comment (davecheney) [inline internal/dag/dag.go]: I think this is not right, but we'll live with it for now.

Fundamentally two services with different ports are different, from the dag point of view. Therefore two services with the same port but different healthcheck or lb strategies are different. 

Either

a. we place those differences on the route object, which matches where they came from in the ingressroute document
b. we make healthcheck params and lb strategy part of the key for the _services map--so extend portmeta to include more parameters.

This is fine to commit for beta.1, please raise an issue to address this for 0.6 final.

Thanks

Comment (stevesloka) [reply internal/dag/dag.go]: If we put this on the Route Object, how would you move that into the CDS update time?

Comment (stevesloka) [reply internal/dag/dag.go]: Also confused about your comment regarding different services, this PR maintains different healthchecks or lb strat per service.

**Comment by stevesloka** [comment]
@davecheney I merged your PR and repushed. Also, please take a look as well as a question on your comment above.

File: internal/dag/dag.go (hunk 8 above)
Comment (davecheney) [reply internal/dag/dag.go]: > Also confused about your comment regarding different services, this PR maintains different healthchecks or lb strat per service.

Sorry, I probably wasn't very clear.

At the moment the _services (terrible name, sorry) map is keyed by a tuple of namespace/name/port; so the same k8s service will result in two different service objects in the dag if a route points to two different ports on the same service.  https://github.com/heptio/contour/blob/master/internal/contour/cluster_test.go#L220

Arguably lbstrat and healthcheck params are properties of the service namespace/name/port tuple. I think I can argue that for the latter, why would the a service have two different ways of checking its health depending on which routes point to it. I could probably make the same argument for lbstrat. If that argument holds then those properties should be defined on the k8s service object via annotations, and incorporated into the dag.Service object -- which means they are _not_ part of the _service's map key.

However, because we currently define lbstrat and hc on the ingressroute object, it is possible to have more than one way to point to a single k8s service with different values for those parameters. So, what we should probably do in that case is make lbstrat and healthcheck params properties of the route so when the cds visitor is building the cds table it knows to emit distinct cluster values for each cluster depending on the path it took to reach that vertex.

However, this argument is problematic because if a route has multiple service children, does that mean we need to store an array of healthcheck and lbstrat configurations on the Route object? The ingressroute document lets us define a different hc and lbstrat per service so we need to store and map an array of hc/lbstrat values in the route object to the route objects children. 

This is the same situation I found myself in when doing RDS, a ingressroute Route can have multiple service name / port pairs. But the child of a dag.Route was a k8s Service object, which can have multiple ports defined. Said another way, the ingressroute Route servicename field gave us the k8s service object to wrap in a dag.Service and insert into the graph, but it didn't contain enough information for the routeVisitor to recover the port details of the child dag.Service. The first solution I tried was to create a map of ports to children in the route object, but that was brittle. The better solution was to create a distinct dag.Service for a service name and port combination. This also matched the envoy Cluster model.

In summary, what is here will work for beta.1, but we need to revisit it before 0.6 final because I think that these parameters belong on the service document, which is problematic for the discoverers.

**Comment by davecheney** [comment]
I'm merging this on @stevesloka 's behalf as its july 4th in the states. I've raised #503 to track the changes needed to the dag or service annotations.
